### PR TITLE
[Dubbo-2131] Check if we can switch serialization/deser between IO and Dubbo Threads.

### PR DESCRIPTION
**Pull request by kid1994**
## What is the purpose of the change

add ability to switch serialization/deser between IO and Dubbo Threads.

## Brief changelog

change dubbo xml schema
add one "decodeInIo" property to dubbo-config/dubbo-config-api/src/main/java/com/alibaba/dubbo/config/ProtocolConfig.java
change "DECODE_IN_IO_THREAD_KEY " to "decodeinio"

## Verifying this change

write an unit-test and passed

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-rpc/dubbo-rpc-dubbo/src/main/java/com/alibaba/dubbo/rpc/protocol/dubbo/DubboCodec.java [hunk 1]
```diff
@@ -90,6 +90,7 @@ protected Object decodeBody(Channel channel, InputStream is, byte[] header) thro
                             result = new DecodeableRpcResult(channel, res, is,
                                     (Invocation) getRequestData(id), proto);
                             result.decode();
+                            System.out.println(Thread.currentThread().getName());
```
Comment (chickenlj) [inline dubbo-rpc/dubbo-rpc-dubbo/src/main/java/com/alibaba/dubbo/rpc/protocol/dubbo/DubboCodec.java]: Remove this line

File: dubbo-common/src/main/java/com/alibaba/dubbo/common/Constants.java [hunk 2]
```diff
@@ -589,7 +589,8 @@
 
     public static final long DEFAULT_TPS_LIMIT_INTERVAL = 60 * 1000;
 
-    public static final String DECODE_IN_IO_THREAD_KEY = "decode.in.io";
+    //public static final String DECODE_IN_IO_THREAD_KEY = "decode.in.io";
+    public static final String DECODE_IN_IO_THREAD_KEY = "decodeinio";
```
Comment (chickenlj) [inline dubbo-common/src/main/java/com/alibaba/dubbo/common/Constants.java]: This change will cause the incompatible problem for users already used this property.

Use @Parameter annotation to solve this problem, see the comment below.

File: dubbo-config/dubbo-config-api/src/main/java/com/alibaba/dubbo/config/ProtocolConfig.java [hunk 3]
```diff
@@ -454,6 +456,14 @@ public void setExtension(String extension) {
         this.extension = extension;
     }
 
+    public Boolean getDecodeInIo() {
```
Comment (chickenlj) [inline dubbo-config/dubbo-config-api/src/main/java/com/alibaba/dubbo/config/ProtocolConfig.java]: Add @Parameter(key = "decode.in.io")

Comment (kid1994) [reply dubbo-config/dubbo-config-api/src/main/java/com/alibaba/dubbo/config/ProtocolConfig.java]: thank you for your repley.there is another group who use the similiar way,and it seems they  dont have to change the src code.I think their way is better.

**Comment by chickenlj** [comment]
1. If we put this property on ProtocolConfig, then it's only available on Provider side.
2. Avoid Provider side configurations affect that on the Consumer side.

Currently, this demo works. If you fixed the above problems, feel free to reopen it. Thanks.

**Pull request closed without merging**